### PR TITLE
Fix: placeholder width cutting off text #291

### LINKED ISSUE #291: Placeholder width cutting off text when Select2 is initialized inside a hidden element
We're having to patch select2 like so to get our placeholder text to appear properly.
- ln 345: width: auto !important;

This happens when the new width option is either not provided or is set to 'element'. The placeholder gets an inline css width set, which mysteriously goes away and never comes back when the input is interacted with.

### PR DESCRIPTION
**Pull request by bews**
This pull request includes a

- [x] Bug fix
- [ ] New feature
- [ ] Translation

Fixes #291, #3669 and other duplicates.

### CODE REVIEW

**Comment by jpic** [comment]
The width change makes sense but is there any risk that the closest call fails to pick the right element ?

If both this.$search and its .select2-search--inline are children of the widget's parent element then it would be safe.

I wonder what @kevin-brown meant in the comment that there is no clean fix here  https://github.com/select2/select2/issues/3669#issuecomment-130096418

**Comment by bews** [comment]
I didn't look deep into the code, but I believe `.select2-search--inline` is always there. And it's the direct parent of `this.$search`.

File: src/js/select2/selection/search.js [hunk 1]
```diff
@@ -206,16 +206,19 @@ define([
...
+      widthParent = '100%';
     } else {
       var minimumWidth = this.$search.val().length + 1;
-
       width = (minimumWidth * 0.75) + 'em';
+      widthParent = 'auto';
     }
 
     this.$search.css('width', width);
+    this.$search.closest('.select2-search--inline').css('width', widthParent);
```
Comment (bews) [inline src/js/select2/selection/search.js]: This can be replaced by `.parent()` to be sure.

Comment (alexweissman) [reply src/js/select2/selection/search.js]: Do you still want to make that change?

**Comment by cssagogo** [comment]
When is this fix going out?

File: src/js/select2/selection/search.js (hunk 1 above)
Comment (bews) [reply src/js/select2/selection/search.js]: It works fine the way it is, but I can change it if you think `.parent()` is a better approach. I'm ok with either.

**Comment by alexweissman** [comment]
It just seems suspicious to me that replacing a calculated value like `width = this.$selection.find('.select2-selection__rendered').innerWidth();` with a hardcoded value like `100%` would be a fix.

Surely there must be a reason that it wasn't set to 100% in the first place?

By the way, a more conservative patch is suggested in #3292:

`width = this.$selection.find('.select2-selection__rendered').innerWidth() || '100%';`

**Comment by bews** [comment]
> `width = this.$selection.find('.select2-selection__rendered').innerWidth() || '100%';`

This doesn't work, since innerWidth returns incorrect value, not 0. It may be bigger than it should, or a fixed 100px.

I dunno why it returns that and don't have time to figure out.

**Comment by ala-musleh** [comment]
this css works perfect for me (multiple selection, select starts hidden):

.select2-container .select2-selection__rendered > *:first-child.select2-search--inline {
width: 100% !important;
}
.select2-container .select2-selection__rendered > *:first-child.select2-search--inline .select2-search__field {
width: 100% !important;
}

**Comment by vpn** [comment]
Thanks for that, @ala-musleh.

**Comment by spo0okie** [comment]
this problem stil exists on visible elements (not only for hidden ones from closed issues) for multiple select, and that fix works perfect. why it sitll not merged?

**Comment by seaexplorer2024** [comment]
just commenting I still get the same issue on V4.0.13 for multi selects.

Using @xander-mu's styling has fixed it for me

**Comment by DonSYS91** [comment]
Issue still exists as of now as well on latest version.
Thank you so much @xander-mu  for the workaround

**Pull request closed without merging**